### PR TITLE
KAFKA-7853: Refactor coordinator config

### PR DESCRIPTION
**Pull request by abbccdda**
A test attempt to refactor current coordinator logic.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by abbccdda** [comment]
@stanislavkozlovski @hachikuji Do you think this makes sense to you?

File: clients/src/main/java/org/apache/kafka/clients/GroupRebalanceConfig.java
Comment (stanislavkozlovski) [inline clients/src/main/java/org/apache/kafka/clients/GroupRebalanceConfig.java:28]: Should we call this something like `ConsumerGroupConfig`? All the configs here aren't directly linked to rebalancing

**Review by stanislavkozlovski — COMMENTED** [review]
I think these changes make a lot of sense, we were discussing how the Coordinator constructor was becoming too bloated, so I'm glad we're getting to shave it off :)

**Comment by guozhangwang** [comment]
retest this please

File: clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java:104]: Why do we want to move consumer-only configs to common-client configs?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: nit: better name variable to `rebalanceConfig`. Class name is good though.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MockGroupRebalanceConfig.java
Comment (guozhangwang) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MockGroupRebalanceConfig.java]: This mock seems not necessary as it was not used anywhere.

And even in some unit test cases, I think we can just use a dummy config?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java
Comment (guozhangwang) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java]: This is not used.

**Review by guozhangwang — COMMENTED** [review]
The PR lgtm but still seems it is not yet ready for reviews as some cleanup are not done (left some comments).

File: clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java
Comment (abbccdda) [reply clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java:104]: `group.id` is not consumer-only. It is also used in `DistributedConfig`, and if we want to consolidate all the rebalance related configs, we need to have it defined in common area.

File: clients/src/main/java/org/apache/kafka/clients/GroupRebalanceConfig.java
Comment (abbccdda) [reply clients/src/main/java/org/apache/kafka/clients/GroupRebalanceConfig.java:28]: Yea, I thought about that too. However, these configs are not constrained to KConsumer either. Connect also uses them.

File: clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java:104]: Sounds good to me.

**Comment by guozhangwang** [comment]
I've made another pass on the PR and it looks good to me. Would like to ping another pair of eyes to review before merging, cc @stanislavkozlovski @hachikuji @rajinisivaram.

Also, could you rebase (there's a minor conflict).

File: clients/src/main/java/org/apache/kafka/clients/GroupRebalanceConfig.java
Comment (kkonstantine) [inline clients/src/main/java/org/apache/kafka/clients/GroupRebalanceConfig.java]: can be `final`

File: clients/src/main/java/org/apache/kafka/clients/GroupRebalanceConfig.java
Comment (kkonstantine) [inline clients/src/main/java/org/apache/kafka/clients/GroupRebalanceConfig.java]: usually overriding toString is preferable for enum types
We also don't use `get/set` a lot in the codebase and rather use names that corresponds to the return values.

File: clients/src/main/java/org/apache/kafka/clients/GroupRebalanceConfig.java
Comment (kkonstantine) [inline clients/src/main/java/org/apache/kafka/clients/GroupRebalanceConfig.java]: I think it's a bit unusual not to have the enum name all caps. 
I'd images it would be CONSUMER, CONNECT

File: clients/src/main/java/org/apache/kafka/clients/GroupRebalanceConfig.java
Comment (kkonstantine) [inline clients/src/main/java/org/apache/kafka/clients/GroupRebalanceConfig.java]: Also if you intend to have a 1-1 mapping between the enum name and the `String` that is returned here, you may as well skip the field and use:
`return name().toLowerCase(Locale.ROOT);`

Comment (kkonstantine) [reply clients/src/main/java/org/apache/kafka/clients/GroupRebalanceConfig.java]: Some enum types do that (e.g. [ConnectorType](https://github.com/apache/kafka/blob/trunk/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/entities/ConnectorType.java)). Some others use a `private final` field and the constructor ([TimestampType](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/record/TimestampType.java)). Up to you

File: clients/src/main/java/org/apache/kafka/clients/GroupRebalanceConfig.java
Comment (kkonstantine) [inline clients/src/main/java/org/apache/kafka/clients/GroupRebalanceConfig.java:78]: There is a variable `ConsumerConfig#LEAVE_GROUP_ON_CLOSE_CONFIG` for this property

**Review by kkonstantine — COMMENTED** [review]
Thanks @abbccdda 
I left a few comments related to the new code. The rest is refactoring IIUC.

File: clients/src/main/java/org/apache/kafka/clients/GroupRebalanceConfig.java
Comment (abbccdda) [reply clients/src/main/java/org/apache/kafka/clients/GroupRebalanceConfig.java]: Sounds good!

File: clients/src/main/java/org/apache/kafka/clients/GroupRebalanceConfig.java
Comment (abbccdda) [reply clients/src/main/java/org/apache/kafka/clients/GroupRebalanceConfig.java]: I would prefer first one for its cleanness

File: clients/src/main/java/org/apache/kafka/clients/GroupRebalanceConfig.java
Comment (abbccdda) [reply clients/src/main/java/org/apache/kafka/clients/GroupRebalanceConfig.java:78]: Yea, unfortunately it's a private config which we couldn't take out.

**Comment by abbccdda** [comment]
System tests passed for consumer and connect: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2718/ and https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2716/

File: clients/src/main/java/org/apache/kafka/clients/GroupRebalanceConfig.java
Comment (kkonstantine) [reply clients/src/main/java/org/apache/kafka/clients/GroupRebalanceConfig.java:78]: Got it. Ok, makes sense for now I guess.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java
Comment (kkonstantine) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java]: Related to my other comment below, we change who we initialize `groupInstanceId`, but this seems to result into more manual initialization within the tests than fewer (as someone would expect by the use of a `setup` method). Should we keep the current way of initialization?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java
Comment (kkonstantine) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java]: In many test cases now we don't seem to take advantage of initialization of `coordinator` in the `setup` method. Any reason why this changed?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java
Comment (kkonstantine) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java]: Also these declarations are masking member fields. We should avoid that because it might lead to bugs that are hard to detect. If we have member fields for any of the variables we need to customize in a test case, we should use those and avoid masking with local variables.

File: clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java
Comment (kkonstantine) [inline clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java]: nit: since we are changing this, probably should be one of the following: 
`Configurations shared by Kafka client applications: producer, consumer, connect, etc.`
or 
`Configurations shared by Kafka client applications: producer, consumer, connect and adminclient.`

adminclient is one that is missing, so the version with `etc` might be preferable.

**Review by kkonstantine — COMMENTED** [review]
Thanks @abbccdda for addressing the latest comments. 
Took another pass and left a few minor comments mostly in tests.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java
Comment (abbccdda) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java]: Sounds good, let me remove some of them

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java
Comment (abbccdda) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java]: `testLeaveGroupOnClose`, `testPendingMemberShouldLeaveGroup` and `testMaybeLeaveGroup` need reinitialization because we need to make sure the rebalance config points to empty `group.instance.id` id so that `leaveGroupOnClose` will be false.

Comment (abbccdda) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java]: Nvm, I think there are not needed.

File: clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java
Comment (abbccdda) [reply clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java]: Sounds good!

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java
Comment (abbccdda) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java]: +1

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java
Comment (abbccdda) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java]: Sounds good, default to dynamic membership setup

**Review by kkonstantine — APPROVED** [review]
Thanks for addressing all the comments @abbccdda 
This LGTM!

**Pull request merged**
